<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Create from fastq &#8212; loompy 3.0.6 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="API Walkthrough" href="../apiwalkthrough/index.html" />
    <link rel="prev" title="Understanding the semantics of loom files" href="../semantics/index.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="../_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="../_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-light navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html"><span><img src="../_static/Loom_icon.png"></span>
          Loom</a>
        <span class="navbar-text navbar-version pull-left"><b>3.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="http://loompy.org">loompy.org</a></li>
                <li><a href="https://github.com/linnarsson-lab/loompy">GitHub</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Contents <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Loompy documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation/index.html#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/index.html#easy-installation">Easy installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/index.html#from-source">From source</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/index.html#getting-started">Getting Started</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../semantics/index.html">Understanding the semantics of loom files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../semantics/index.html#connecting-not-loading-and-saving">Connecting, not loading and saving</a></li>
<li class="toctree-l2"><a class="reference internal" href="../semantics/index.html#reading-and-writing">Reading and writing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../semantics/index.html#efficient-indexing-and-compression">Efficient indexing and compression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../semantics/index.html#matrix-and-attributes">Matrix and attributes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../semantics/index.html#data-types">Data types</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Create from fastq</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#the-loompy-command-line-tool">The <code class="docutils literal notranslate"><span class="pre">loompy</span></code> command-line tool</a></li>
<li class="toctree-l2"><a class="reference internal" href="#using-the-loompy-fromfq-command">Using the <code class="docutils literal notranslate"><span class="pre">loompy</span> <span class="pre">fromfq</span></code> command</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../apiwalkthrough/index.html">API Walkthrough</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../apiwalkthrough/index.html#creating-and-connecting">Creating and connecting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../apiwalkthrough/index.html#manipulate-data">Manipulate data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cookbook/index.html">Cookbook</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../cookbook/index.html#working-with-a-newly-created-file">Working with a newly created file</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cookbook/index.html#loading-attributes-from-pandas">Loading attributes from Pandas</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cookbook/index.html#combining-data-using-scan-and-new">Combining data using scan() and new()</a></li>
<li class="toctree-l2"><a class="reference internal" href="../cookbook/index.html#fitting-an-incremental-pca">Fitting an incremental PCA</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../conventions/index.html">Conventions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../conventions/index.html#single-analysis-per-file">Single analysis per file</a></li>
<li class="toctree-l2"><a class="reference internal" href="../conventions/index.html#orientation">Orientation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../conventions/index.html#attribute-naming-conventions">Attribute naming conventions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../format/index.html">Loom file format specs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../format/index.html#versions">Versions</a></li>
<li class="toctree-l2"><a class="reference internal" href="../format/index.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../format/index.html#hdf5-concepts">HDF5 concepts</a></li>
<li class="toctree-l2"><a class="reference internal" href="../format/index.html#standards">Standards</a></li>
<li class="toctree-l2"><a class="reference internal" href="../format/index.html#specification">Specification</a></li>
<li class="toctree-l2"><a class="reference internal" href="../format/index.html#datatypes">Datatypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../format/index.html#example">Example</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../fullapi/index.html">Complete API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/create.html">create (function)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/new.html">new (function)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/create_from_fastq.html">create_from_fastq (function)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/create_from_cellranger.html">create_from_cellranger (function)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/connect.html">connect (function)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/combine.html">combine (function)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/loompy.html">LoomConnection (class)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/layer_manager.html">LayerManager (class)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/memory_loom_layer.html">MemoryLoomLayer (class)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/loom_layer.html">LoomLayer (class)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/view_manager.html">ViewManager (class)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/loom_view.html">LoomView (class)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fullapi/loom_validator.html">LoomValidator (class)</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><div class="list-group">
    <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Loompy documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../semantics/index.html">Understanding the semantics of loom files</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Create from fastq</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#the-loompy-command-line-tool">The <code class="docutils literal notranslate"><span class="pre">loompy</span></code> command-line tool</a></li>
<li class="toctree-l2"><a class="reference internal" href="#using-the-loompy-fromfq-command">Using the <code class="docutils literal notranslate"><span class="pre">loompy</span> <span class="pre">fromfq</span></code> command</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#install-kallisto">1. Install kallisto</a></li>
<li class="toctree-l3"><a class="reference internal" href="#download-an-index-or-build-your-own">2. Download an index (or build your own)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#download-metadata-and-fastq-files">3. Download metadata and fastq files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#run-the-loompy-fromfq-command">4. Run the <code class="docutils literal notranslate"><span class="pre">loompy</span> <span class="pre">fromfq</span></code> command</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../apiwalkthrough/index.html">API Walkthrough</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cookbook/index.html">Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conventions/index.html">Conventions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../format/index.html">Loom file format specs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../fullapi/index.html">Complete API Reference</a></li>
</ul>

</div>
        </div>
      </div>
    <div class="body col-md-9 content" role="main">
      
  <div class="section" id="create-from-fastq">
<span id="kallisto"></span><h1>Create from fastq<a class="headerlink" href="#create-from-fastq" title="Permalink to this headline">¶</a></h1>
<p>This section introduces the <code class="docutils literal notranslate"><span class="pre">loompy</span></code> command-line tool, and shows how it is used to create .loom files directly from fastq files.</p>
<div class="section" id="the-loompy-command-line-tool">
<h2>The <code class="docutils literal notranslate"><span class="pre">loompy</span></code> command-line tool<a class="headerlink" href="#the-loompy-command-line-tool" title="Permalink to this headline">¶</a></h2>
<p>After installing loompy 3, a new command-line tool <code class="docutils literal notranslate"><span class="pre">loompy</span></code> will be available. To verify, launch a Terminal window and type <code class="docutils literal notranslate"><span class="pre">loompy</span></code>,
and you should see the following output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Usage</span><span class="p">:</span> <span class="n">loompy</span> <span class="p">[</span><span class="n">OPTIONS</span><span class="p">]</span> <span class="n">COMMAND</span> <span class="p">[</span><span class="n">ARGS</span><span class="p">]</span><span class="o">...</span>

<span class="n">Options</span><span class="p">:</span>
  <span class="o">--</span><span class="n">show</span><span class="o">-</span><span class="n">message</span> <span class="o">/</span> <span class="o">--</span><span class="n">hide</span><span class="o">-</span><span class="n">message</span>
  <span class="o">--</span><span class="n">verbosity</span> <span class="p">[</span><span class="n">error</span><span class="o">|</span><span class="n">warning</span><span class="o">|</span><span class="n">info</span><span class="o">|</span><span class="n">debug</span><span class="p">]</span>
  <span class="o">--</span><span class="n">help</span>                          <span class="n">Show</span> <span class="n">this</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span><span class="o">.</span>

<span class="n">Commands</span><span class="p">:</span>
  <span class="n">fromfq</span>
</pre></div>
</div>
</div>
<div class="section" id="using-the-loompy-fromfq-command">
<h2>Using the <code class="docutils literal notranslate"><span class="pre">loompy</span> <span class="pre">fromfq</span></code> command<a class="headerlink" href="#using-the-loompy-fromfq-command" title="Permalink to this headline">¶</a></h2>
<div class="section" id="install-kallisto">
<h3>1. Install <a class="reference external" href="https://pachterlab.github.io/kallisto/">kallisto</a><a class="headerlink" href="#install-kallisto" title="Permalink to this headline">¶</a></h3>
<p>The excellent kallisto tool performs ultra-fast pseudoalignment, which loompy uses to count reads (UMIs) on genes.</p>
</div>
<div class="section" id="download-an-index-or-build-your-own">
<h3>2. Download an index (or build your own)<a class="headerlink" href="#download-an-index-or-build-your-own" title="Permalink to this headline">¶</a></h3>
<p>We provide a pre-built index of the <a class="reference external" href="https://storage.googleapis.com/linnarsson-lab-www-blobs/human_GRCh38_gencode.v31.tar.gz">human genome</a>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This index is really only suitable for use with 10x Chromium data (it makes strong assumptions about the read distribution).</p>
</div>
<p>Unzip the index to a directory, which will have the following content:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">10</span><span class="n">xv1_whitelist</span><span class="o">.</span><span class="n">txt</span>
<span class="mi">10</span><span class="n">xv2_whitelist</span><span class="o">.</span><span class="n">txt</span>
<span class="mi">10</span><span class="n">xv3_whitelist</span><span class="o">.</span><span class="n">txt</span>
<span class="n">fragments2genes</span><span class="o">.</span><span class="n">txt</span>
<span class="n">gencode</span><span class="o">.</span><span class="n">v31</span><span class="o">.</span><span class="n">fragments</span><span class="o">.</span><span class="n">idx</span>
<span class="n">gencode</span><span class="o">.</span><span class="n">v31</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">tab</span>
<span class="n">manifest</span><span class="o">.</span><span class="n">json</span>
<span class="n">spliced_fragments</span><span class="o">.</span><span class="n">txt</span>
<span class="n">unspliced_fragments</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>To build your own index, see the <a class="reference external" href="https://github.com/linnarsson-lab/loompy/blob/master/notebooks/build_index.ipynb">build instructions</a>.</p>
</div>
<div class="section" id="download-metadata-and-fastq-files">
<h3>3. Download metadata and fastq files<a class="headerlink" href="#download-metadata-and-fastq-files" title="Permalink to this headline">¶</a></h3>
<p>You need to provide the input fastq files, and metadata for your samples. For this tutorial, please download the 1k PBMC v3 reference dataset
from 10x Genomics: <a class="reference external" href="http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_fastqs.tar">pbmc_1k_v3_fastqs.tar</a>.</p>
<p>Metadata can be provided either in the form of a tab-delimited file with a single header row, or as a sqlite3 database with a table named <code class="docutils literal notranslate"><span class="pre">sample</span></code>.</p>
<p>One metadata column must be named <code class="docutils literal notranslate"><span class="pre">name</span></code> and contain sample names (sample IDs). One column must be named <code class="docutils literal notranslate"><span class="pre">technology</span></code> and contain the technology name (currently one of <code class="docutils literal notranslate"><span class="pre">10xv1</span></code>,
<code class="docutils literal notranslate"><span class="pre">10xv2</span></code> or <code class="docutils literal notranslate"><span class="pre">10xv3</span></code>). The technology name will be used to locate the barcode whitelist file (e.g. <code class="docutils literal notranslate"><span class="pre">10xv3_whitelist.txt</span></code>). Finally, one
column must be named <code class="docutils literal notranslate"><span class="pre">targetnumcells</span></code> and contain the target (or expected) cell number in the samples. If absent, the default will be 5000.</p>
<p>The metadata file (or database) will typically contain metadata for all the samples you are working with, one row per sample. You can add any number of
additional columns of metadata, which will all be copied to the newly generated loom file as global attributes.</p>
<p>For this tutorial, create a tab-delimited file named <code class="docutils literal notranslate"><span class="pre">metadata.tab</span></code> with the following content:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">name</span>    <span class="n">technology</span>  <span class="n">targetnumcells</span>
<span class="mi">1</span><span class="n">kPBMC</span>  <span class="mi">10</span><span class="n">xv3</span>       <span class="mi">1000</span>
</pre></div>
</div>
<p>(If you copy-paste the above, make sure the fields are tab delimited)</p>
</div>
<div class="section" id="run-the-loompy-fromfq-command">
<h3>4. Run the <code class="docutils literal notranslate"><span class="pre">loompy</span> <span class="pre">fromfq</span></code> command<a class="headerlink" href="#run-the-loompy-fromfq-command" title="Permalink to this headline">¶</a></h3>
<p>Run the following command (replace <code class="docutils literal notranslate"><span class="pre">human_GRCh38_gencode.v31</span></code> with the path to the human genome index you just downloaded):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">loompy</span> <span class="n">fromfq</span> <span class="mi">1</span><span class="n">kPBMC</span><span class="o">.</span><span class="n">loom</span> <span class="mi">1</span><span class="n">kPBMC</span> <span class="n">human_GRCh38_gencode</span><span class="o">.</span><span class="n">v31</span> <span class="n">metadata</span><span class="o">.</span><span class="n">tab</span> <span class="n">pbmc_1k_v3_S1_L001_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">pbmc_1k_v3_S1_L001_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">pbmc_1k_v3_S1_L002_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">pbmc_1k_v3_S1_L002_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Note that the fastq files are listed in pairs of R1 (read 1) and R2 (read 2) files. The index files (I1) are not used.</p>
<p>After about half an hour, you will have a <code class="docutils literal notranslate"><span class="pre">1kPBMC.loom</span></code> file with separate <code class="docutils literal notranslate"><span class="pre">spliced</span></code> and <code class="docutils literal notranslate"><span class="pre">unspliced</span></code> layers (the main matrix will be
the sum of the two), and rich metadata for both genes, cells and the sample itself stored as attributes. The output should look something like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">08</span><span class="p">,</span><span class="mi">186</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">kallisto</span> <span class="n">bus</span> <span class="o">-</span><span class="n">i</span> <span class="n">human_GRCh38_gencode</span><span class="o">.</span><span class="n">v31</span><span class="o">/</span><span class="n">gencode</span><span class="o">.</span><span class="n">v31</span><span class="o">.</span><span class="n">fragments</span><span class="o">.</span><span class="n">idx</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">tmp</span><span class="o">/</span><span class="n">tmp7yk3rf07</span> <span class="o">-</span><span class="n">x</span> <span class="mi">10</span><span class="n">xv3</span> <span class="o">-</span><span class="n">t</span> <span class="mi">56</span> <span class="n">pbmc_1k_v3_S1_L001_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">pbmc_1k_v3_S1_L001_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">pbmc_1k_v3_S1_L002_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="n">pbmc_1k_v3_S1_L002_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">08</span><span class="p">,</span><span class="mi">307</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="n">k</span><span class="o">-</span><span class="n">mer</span> <span class="n">length</span><span class="p">:</span> <span class="mi">31</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">08</span><span class="p">,</span><span class="mi">307</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="n">number</span> <span class="n">of</span> <span class="n">targets</span><span class="p">:</span> <span class="mi">845</span><span class="p">,</span><span class="mi">338</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">08</span><span class="p">,</span><span class="mi">307</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="n">number</span> <span class="n">of</span> <span class="n">k</span><span class="o">-</span><span class="n">mers</span><span class="p">:</span> <span class="mi">178</span><span class="p">,</span><span class="mi">605</span><span class="p">,</span><span class="mi">364</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">29</span><span class="p">,</span><span class="mi">537</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="n">number</span> <span class="n">of</span> <span class="n">equivalence</span> <span class="n">classes</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span><span class="mi">191</span><span class="p">,</span><span class="mi">221</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span><span class="mi">951</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="p">[</span><span class="n">quant</span><span class="p">]</span> <span class="n">will</span> <span class="n">process</span> <span class="n">sample</span> <span class="mi">1</span><span class="p">:</span> <span class="n">pbmc_1k_v3_S1_L001_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span><span class="mi">951</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span>                                <span class="n">pbmc_1k_v3_S1_L001_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span><span class="mi">951</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="p">[</span><span class="n">quant</span><span class="p">]</span> <span class="n">will</span> <span class="n">process</span> <span class="n">sample</span> <span class="mi">2</span><span class="p">:</span> <span class="n">pbmc_1k_v3_S1_L002_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">28</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span><span class="mi">951</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span>                                <span class="n">pbmc_1k_v3_S1_L002_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">31</span><span class="p">:</span><span class="mi">44</span><span class="p">,</span><span class="mi">144</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="p">[</span><span class="n">quant</span><span class="p">]</span> <span class="n">finding</span> <span class="n">pseudoalignments</span> <span class="k">for</span> <span class="n">the</span> <span class="n">reads</span> <span class="o">...</span> <span class="n">done</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">31</span><span class="p">:</span><span class="mi">44</span><span class="p">,</span><span class="mi">145</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="p">[</span><span class="n">quant</span><span class="p">]</span> <span class="n">processed</span> <span class="mi">66</span><span class="p">,</span><span class="mi">601</span><span class="p">,</span><span class="mi">887</span> <span class="n">reads</span><span class="p">,</span> <span class="mi">46</span><span class="p">,</span><span class="mi">119</span><span class="p">,</span><span class="mi">840</span> <span class="n">reads</span> <span class="n">pseudoaligned</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">31</span><span class="p">:</span><span class="mi">52</span><span class="p">,</span><span class="mi">543</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Loading</span> <span class="n">gene</span> <span class="n">metadata</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">31</span><span class="p">:</span><span class="mi">52</span><span class="p">,</span><span class="mi">818</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Loading</span> <span class="n">fragments</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">gene</span> <span class="n">mappings</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">31</span><span class="p">:</span><span class="mi">53</span><span class="p">,</span><span class="mi">426</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Indexing</span> <span class="n">genes</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">31</span><span class="p">:</span><span class="mi">53</span><span class="p">,</span><span class="mi">846</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Loading</span> <span class="n">equivalence</span> <span class="n">classes</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">32</span><span class="p">:</span><span class="mi">22</span><span class="p">,</span><span class="mi">273</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Mapping</span> <span class="n">equivalence</span> <span class="n">classes</span> <span class="n">to</span> <span class="n">genes</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">32</span><span class="p">:</span><span class="mi">32</span><span class="p">,</span><span class="mi">817</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Loading</span> <span class="n">fragment</span> <span class="n">IDs</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">32</span><span class="p">:</span><span class="mi">33</span><span class="p">,</span><span class="mi">280</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Loading</span> <span class="n">BUS</span> <span class="n">records</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">33</span><span class="p">:</span><span class="mi">46</span><span class="p">,</span><span class="mi">692</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Sorting</span> <span class="n">cell</span> <span class="n">IDs</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">33</span><span class="p">:</span><span class="mi">49</span><span class="p">,</span><span class="mi">611</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Found</span> <span class="mi">46</span><span class="p">,</span><span class="mi">119</span><span class="p">,</span><span class="mi">840</span> <span class="n">records</span> <span class="k">for</span> <span class="mi">60</span><span class="p">,</span><span class="mi">662</span> <span class="n">genes</span> <span class="ow">and</span> <span class="mi">551</span><span class="p">,</span><span class="mi">892</span> <span class="n">uncorrected</span> <span class="n">cell</span> <span class="n">barcodes</span><span class="o">.</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">33</span><span class="p">:</span><span class="mi">49</span><span class="p">,</span><span class="mi">611</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Correcting</span> <span class="n">cell</span> <span class="n">barcodes</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">35</span><span class="p">:</span><span class="mi">58</span><span class="p">,</span><span class="mi">753</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Found</span> <span class="mi">307</span><span class="p">,</span><span class="mi">677</span> <span class="n">corrected</span> <span class="n">cell</span> <span class="n">barcodes</span><span class="o">.</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">35</span><span class="p">:</span><span class="mi">58</span><span class="p">,</span><span class="mi">754</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Removing</span> <span class="n">redundant</span> <span class="n">reads</span> <span class="n">using</span> <span class="n">UMIs</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">36</span><span class="p">:</span><span class="mi">45</span><span class="p">,</span><span class="mi">546</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="mi">71</span><span class="o">%</span> <span class="n">sequencing</span> <span class="n">saturation</span><span class="o">.</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">36</span><span class="p">:</span><span class="mi">45</span><span class="p">,</span><span class="mi">546</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Counting</span> <span class="n">pseudoalignments</span> <span class="k">for</span> <span class="n">main</span> <span class="n">matrix</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">36</span><span class="p">:</span><span class="mi">52</span><span class="p">,</span><span class="mi">752</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Found</span> <span class="mi">5</span><span class="p">,</span><span class="mi">027</span><span class="p">,</span><span class="mi">188</span> <span class="n">UMIs</span><span class="o">.</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">36</span><span class="p">:</span><span class="mi">53</span><span class="p">,</span><span class="mi">536</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Counting</span> <span class="n">pseudoalignments</span> <span class="k">for</span> <span class="n">layer</span> <span class="s1">&#39;unspliced&#39;</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">38</span><span class="p">:</span><span class="mi">00</span><span class="p">,</span><span class="mi">099</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Found</span> <span class="mi">2</span><span class="p">,</span><span class="mi">376</span><span class="p">,</span><span class="mi">590</span> <span class="n">UMIs</span><span class="o">.</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">38</span><span class="p">:</span><span class="mi">00</span><span class="p">,</span><span class="mi">706</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Counting</span> <span class="n">pseudoalignments</span> <span class="k">for</span> <span class="n">layer</span> <span class="s1">&#39;spliced&#39;</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">39</span><span class="p">:</span><span class="mi">09</span><span class="p">,</span><span class="mi">718</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Found</span> <span class="mi">3</span><span class="p">,</span><span class="mi">231</span><span class="p">,</span><span class="mi">999</span> <span class="n">UMIs</span><span class="o">.</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">39</span><span class="p">:</span><span class="mi">09</span><span class="p">,</span><span class="mi">718</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Calling</span> <span class="n">cells</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">42</span><span class="p">:</span><span class="mi">32</span><span class="p">,</span><span class="mi">387</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Found</span> <span class="mi">1189</span> <span class="n">valid</span> <span class="n">cells</span> <span class="ow">and</span> <span class="o">~</span><span class="mi">77</span> <span class="n">ambient</span> <span class="n">UMIs</span><span class="o">.</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">42</span><span class="p">:</span><span class="mi">32</span><span class="p">,</span><span class="mi">388</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Creating</span> <span class="n">loom</span> <span class="n">file</span> <span class="s1">&#39;1kPBMC.loom&#39;</span>
<span class="mi">2019</span><span class="o">-</span><span class="mi">09</span><span class="o">-</span><span class="mi">29</span> <span class="mi">16</span><span class="p">:</span><span class="mi">42</span><span class="p">:</span><span class="mi">32</span><span class="p">,</span><span class="mi">388</span> <span class="o">-</span> <span class="n">INFO</span> <span class="o">-</span> <span class="n">Saving</span>
</pre></div>
</div>
<p>As you can see, 46,119,840 of 66,601,887 reads pseudoaligned (~70%) which is typical. The sequencing saturation was 71%, and the cell
calling algorithm found 1189 valid cells (similar to the 1,222 cells reported by cellranger). Empty beads carried a median of 77
UMIs, presumably from cell-free ambient RNA.</p>
<p>Connect to the loom file and examine its global attributes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">loompy</span>
<span class="k">with</span> <span class="n">loompy</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="s2">&quot;1kPBMC.loom&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ds</span><span class="p">:</span>
  <span class="nb">print</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="p">[</span><span class="s1">&#39;AmbientPValue&#39;</span><span class="p">,</span> <span class="s1">&#39;AmbientUMIs&#39;</span><span class="p">,</span> <span class="s1">&#39;BarcodeTotalUMIs&#39;</span><span class="p">,</span> <span class="s1">&#39;CellBarcodes&#39;</span><span class="p">,</span> <span class="s1">&#39;CreationDate&#39;</span><span class="p">,</span> <span class="s1">&#39;KallistoCommand&#39;</span><span class="p">,</span> <span class="s1">&#39;KallistoVersion&#39;</span><span class="p">,</span> <span class="s1">&#39;LOOM_SPEC_VERSION&#39;</span><span class="p">,</span> <span class="s1">&#39;NumPseudoaligned&#39;</span><span class="p">,</span> <span class="s1">&#39;NumReadsProcessed&#39;</span><span class="p">,</span> <span class="s1">&#39;RedundantReadFraction&#39;</span><span class="p">,</span> <span class="s1">&#39;SampleID&#39;</span><span class="p">,</span> <span class="s1">&#39;Saturation&#39;</span><span class="p">,</span> <span class="s1">&#39;Species&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;targetnumcells&#39;</span><span class="p">,</span> <span class="s1">&#39;technology&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>…column attributes…</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">loompy</span>
<span class="k">with</span> <span class="n">loompy</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="s2">&quot;1kPBMC.loom&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ds</span><span class="p">:</span>
  <span class="nb">print</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">ca</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="p">[</span><span class="s1">&#39;CellID&#39;</span><span class="p">,</span> <span class="s1">&#39;TotalUMIs&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>…row attributes (see the <a class="reference external" href="https://github.com/linnarsson-lab/loompy/blob/master/notebooks/build_index.ipynb">index build instructions</a> for an explanation of these)…</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">loompy</span>
<span class="k">with</span> <span class="n">loompy</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="s2">&quot;1kPBMC.loom&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ds</span><span class="p">:</span>
  <span class="nb">print</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">ra</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="p">[</span><span class="s1">&#39;Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;AccessionVersion&#39;</span><span class="p">,</span> <span class="s1">&#39;Aliases&#39;</span><span class="p">,</span> <span class="s1">&#39;CcdsID&#39;</span><span class="p">,</span> <span class="s1">&#39;Chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;ChromosomeEnd&#39;</span><span class="p">,</span> <span class="s1">&#39;ChromosomeStart&#39;</span><span class="p">,</span> <span class="s1">&#39;CosmicID&#39;</span><span class="p">,</span> <span class="s1">&#39;DnaBindingDomain&#39;</span><span class="p">,</span> <span class="s1">&#39;FullName&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;GeneType&#39;</span><span class="p">,</span> <span class="s1">&#39;HgncID&#39;</span><span class="p">,</span> <span class="s1">&#39;IsTF&#39;</span><span class="p">,</span> <span class="s1">&#39;Location&#39;</span><span class="p">,</span> <span class="s1">&#39;LocationSortable&#39;</span><span class="p">,</span> <span class="s1">&#39;LocusGroup&#39;</span><span class="p">,</span> <span class="s1">&#39;LocusType&#39;</span><span class="p">,</span> <span class="s1">&#39;MgdID&#39;</span><span class="p">,</span> <span class="s1">&#39;MirBaseID&#39;</span><span class="p">,</span> <span class="s1">&#39;OmimID&#39;</span><span class="p">,</span> <span class="s1">&#39;PubmedID&#39;</span><span class="p">,</span> <span class="s1">&#39;RefseqID&#39;</span><span class="p">,</span> <span class="s1">&#39;RgdID&#39;</span><span class="p">,</span> <span class="s1">&#39;UcscID&#39;</span><span class="p">,</span> <span class="s1">&#39;UniprotID&#39;</span><span class="p">,</span> <span class="s1">&#39;VegaID&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>…and layers:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">loompy</span>
<span class="k">with</span> <span class="n">loompy</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="s2">&quot;1kPBMC.loom&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ds</span><span class="p">:</span>
  <span class="nb">print</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">layers</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="p">[</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;spliced&#39;</span><span class="p">,</span> <span class="s1">&#39;unspliced&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2017, LinnarssonLab.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0.<br/>
    </p>
  </div>
</footer>
  </body>
</html>